### PR TITLE
17 add educational resource

### PR DESCRIPTION
**Pull request by antonis96**
Implementation of educational resource entity.

The schema for this entity is  based on **metadata schema** and the **structure of CMS**

### CODE REVIEW

File: src/database/model/educational_resource.py [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+
+
+class EducationalResource(Base):
+    """Any educational resource"""
+
+    __tablename__ = "educational_resources"
+
+    # required fields
+    title: Mapped[str] = mapped_column(String(500), nullable=False)
+    date_modified: Mapped[Date] = mapped_column(DateTime(timezone=True), server_default=func.now())
```
Comment (josvandervelde) [inline src/database/model/educational_resource.py:91]: Is this the date the resource was modified on the original platform (e.g. Zenodo)? Or on our platform? A clarification would be nice, I think. In the former case, the `server_default` should not be there

File: src/database/model/educational_resource.py [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+    body: Mapped[str] = mapped_column(String(500), nullable=False)
+    website_url: Mapped[str] = mapped_column(String(500), nullable=False)
+    educational_role: Mapped[str] = mapped_column(String(500), nullable=False)
+    educational_level: Mapped[str] = mapped_column(String(500), nullable=False)
+    educatonal_type: Mapped[str] = mapped_column(String(500), nullable=False)
```
Comment (josvandervelde) [inline src/database/model/educational_resource.py:96]: typo

File: src/database/model/educational_resource.py [hunk 3]
```diff
@@ -0,0 +1,160 @@
...
+
+    # optional fields
+    interactivity_type: Mapped[str] = mapped_column(String(500), nullable=True)
+    typical_age_range: Mapped[str] = mapped_column(String(500), nullable=True)
+    accessibility_api: Mapped[str] = mapped_column(String(500), nullable=True)
+    accessibility_control: Mapped[str] = mapped_column(String(500), nullable=True)
+    access_mode: Mapped[str] = mapped_column(String(500), nullable=True)
+    access_mode_sufficient: Mapped[str] = mapped_column(String(500), nullable=True)
+    access_restrictions: Mapped[str] = mapped_column(String(500), nullable=True)
+    citation: Mapped[str] = mapped_column(String(500), nullable=True)
```
Comment (josvandervelde) [inline src/database/model/educational_resource.py:107]: This should be an `OrmPublication`, I think. Or even better, a list of them?

File: src/schemas.py [hunk 4]
```diff
@@ -120,3 +99,120 @@ class News(BaseModel):
     business_categories: Optional[list[str]]
     tags: Optional[list[str]]
     id: int | None
+
+
+class DurationMinutesHoursEnum(str, Enum):
```
Comment (josvandervelde) [inline src/schemas.py:104]: I suggest prepending these classes with the prefix `AIoD` --> `AIoDDurationMinutesHoursEnum` etc, to make it easier in the rest of the code to distinguish between the classes

File: src/schemas.py [hunk 5]
```diff
@@ -120,3 +99,120 @@ class News(BaseModel):
...
+    option_1 = "Less than 1 hour"
+    option_2 = "1-3 hours"
+    option_3 = "4-10 hours"
+    option_4 = "10+ hours"
+
+
+class HoursPerWeekEnum(str, Enum):
```
Comment (josvandervelde) [inline src/schemas.py:111]: I think that removing these enums would be better, and just having a string field in the AIoD schema. Then we can have  separate table in the database, more flexible than these enums

File: src/schemas.py [hunk 6]
```diff
@@ -120,3 +99,120 @@ class News(BaseModel):
...
+
+class EducationalResource(BaseModel):
+    """The complete metadata for educational resource"""
+
+    title: str = Field(max_length=500)
+    body: str = Field(max_length=500)
+    website_url: str = Field(max_length=500)
+    date_modified: datetime | None
+
+    educational_role: EducationalRoleEnum
```
Comment (josvandervelde) [inline src/schemas.py:174]: All these fields are now one-on-one. Wouldn't it make sense to have multiple `educational_role`s, or multiple `educational_level`s linked to a single resource? (these are just examples, I think a lot of fields could be repeated)

File: src/schemas.py [hunk 7]
```diff
@@ -120,3 +99,120 @@ class News(BaseModel):
...
+    duration_in_years: int | None
+
+    duration_minutes_and_hours: Optional[DurationMinutesHoursEnum]
+    hours_per_week: Optional[HoursPerWeekEnum]
+    country: Optional[CountryEnum]
+    technical_categories: Optional[list[str]]
+    business_categories: Optional[list[str]]
+    tags: Optional[list[str]]
+
+    id: int | None
```
Comment (josvandervelde) [inline src/schemas.py:218]: Should we add fields `platform` and `platform_identifier` (the identifier given by the platform)? I think that every resource should have them, to make it possible to link it back to the platform it was published on (e.g. Zenodo)

**Review by josvandervelde — CHANGES_REQUESTED** [review]
Looks good! Nice extensive resource

File: src/main.py [hunk 8]
```diff
@@ -566,6 +579,200 @@ def put_news(identifier: str, news: schemas.News) -> dict:
         except Exception as e:
             raise _wrap_as_http_exception(e)
 
+    @app.post(url_prefix + "/educational_resource")
```
[the remark this answers is not in the document]
Comment (antonis96) [reply src/main.py:582]: That looks great. Thanks for this. I am going to refactor it.

File: src/tests/test_get_news.py [hunk 9]
```diff
@@ -58,9 +58,9 @@ def test_happy_path_for_all(client: TestClient, engine: Engine):
     assert {ds["headline"] for ds in response_json} == {"h1", "h2", "h3"}
     assert {ds["source"] for ds in response_json} == {"s1", "s2", "s3"}
     assert {ds["date_modified"] for ds in response_json} == {
-        "2023-03-21",
-        "2023-03-21",
-        "2023-03-21",
+        "2023-03-21T00:00:00",
+        "2023-03-21T00:00:00",
```
[the remark this answers is not in the document]
Comment (antonis96) [reply src/tests/test_get_news.py:62]: fixed [here](https://github.com/aiondemand/AIOD-rest-api/commit/b298737bd5febe8b663e09b0688f5a03f8b0bd26). 
There was an error in expected length of response.
Seems weird since i ran `pre-commit run --all-files` without any problems before opening the PR yesterday.

File: src/database/model/base.py [hunk 10]
```diff
@@ -44,3 +45,44 @@ def to_dict(self, depth: int = 1) -> dict:
             elif value is not None:
                 d[field.name] = value
         return d
+
+
+class BusinessCategory(Base):
```
[the remark this answers is not in the document]
Comment (antonis96) [reply src/database/model/base.py]: fixed [here](https://github.com/aiondemand/AIOD-rest-api/commit/e269573f0c5e31534040a9e8860de237eab15af2)

File: src/database/model/educational_resource.py [hunk 11]
```diff
@@ -0,0 +1,160 @@
...
+    ),
+    Column(
+        "language_id",
+        ForeignKey("languages.id", ondelete="CASCADE", onupdate="RESTRICT"),
+        primary_key=True,
+    ),
+)
+
+
+class EducationalResource(Base):
```
[the remark this answers is not in the document]
Comment (antonis96) [reply src/database/model/educational_resource.py]: The class is more extensive as we try to support CMS structure as well.
Fields that are used can be found [here](https://docs.google.com/spreadsheets/d/1uo5D5lzEU8_xwV3m2TA90Jq5S3NotHpSbI-7zdatInw/edit#gid=1637734839).
We remain in touch with the metadata schema group in order to unify those two spreadsheets.

File: src/database/model/educational_resource.py (hunk 3 above)
Comment (antonis96) [reply src/database/model/educational_resource.py:107]: I used string as a type because of way it is described in the spreadsheet.

`citation | string | A citation or reference to the educational resource. | schema.org`

File: src/schemas.py [hunk 12]
```diff
@@ -120,3 +99,120 @@ class News(BaseModel):
...
+        " such as textual and visual.",
+        default_factory=str,
+    )
+    access_restrictions: str | None = Field(max_length=100)
+    is_accessible_for_free: bool | None
+    time_required: timedelta | None
+    citation: str | None = Field(max_length=200)
+
+    version: str | int | None
+    credits: bool | None
```
[the remark this answers is not in the document]
Comment (antonis96) [reply src/schemas.py:205]: As described in the CMS [spreadsheet](https://docs.google.com/spreadsheets/d/1uo5D5lzEU8_xwV3m2TA90Jq5S3NotHpSbI-7zdatInw/edit#gid=274758270), i understand that we are currently interested only in the existence or not of credits.
In case we want something more, i think your suggestion is the way to go!

File: src/schemas.py (hunk 7 above)
Comment (antonis96) [reply src/schemas.py:218]: Good point! 
Maybe linking **every** resource with a platform must be done in a separate PR.
What do you think on that?

File: src/schemas.py [hunk 13]
```diff
@@ -120,3 +99,120 @@ class News(BaseModel):
...
+    number_of_weeks: int | None
+    field_prerequisites: str | None = Field(max_length=500)
+    short_summary: str | None = Field(max_length=500)
+    duration_in_years: int | None
+
+    duration_minutes_and_hours: Optional[DurationMinutesHoursEnum]
```
[the remark this answers is not in the document]
Comment (antonis96) [reply src/schemas.py:211]: `duration_in_years` is a field mentioned in CMS [spreadsheet](https://docs.google.com/spreadsheets/d/1uo5D5lzEU8_xwV3m2TA90Jq5S3NotHpSbI-7zdatInw/edit#gid=274758270). 
From my point of view it is equivalent to `time_required` so i would suggest keeping the latter one.

Also, `duration_minutes_and_hours` has to be the actual duration of an activity related with the educational resource (maybe a class or something similar).

To sum up, i suggest keeping `time_required` and `duration_minutes_and_hours`

File: src/schemas.py (hunk 6 above)
Comment (antonis96) [reply src/schemas.py:174]: As i can tell from the CMS [spreadsheet](https://docs.google.com/spreadsheets/d/1uo5D5lzEU8_xwV3m2TA90Jq5S3NotHpSbI-7zdatInw/edit#gid=274758270) `educational_level` must only have one value.

Also, `educational_role` seems to be the same as `TargetAudience`  where a `1:n` relationship is used

File: src/database/model/educational_resource.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (josvandervelde) [reply src/database/model/educational_resource.py]: So for now we should see the CMS spreadsheet as the source of truth? This nullifies most of my comments on the data structure. Would it not be easier to first create an AIoD scheme? Now it's conceivable that there will be quite some changes lateron.

File: src/schemas.py (hunk 7 above)
Comment (josvandervelde) [reply src/schemas.py:218]: That's fine with me, although it's not a lot of work (adding two fields), and most other resources already have the fields, I think. Also, it is necessary when you're using the new `ResourceRouter`, because it adds endpoints such as `nodes/{node}/educational_resource/{identifier}`, which uses the `platform` (still called `node` here) and `platform_identifier` (still called `node_specific_identifier` here). So I think it's easier to do it directly, but it's also fine to do it later!

File: src/schemas.py (hunk 5 above)
Comment (antonis96) [reply src/schemas.py]: Enumerations are now gone and replaced with str fields in the AIoD schema as suggested.

File: src/schemas.py (hunk 4 above)
Comment (antonis96) [reply src/schemas.py]: fixed [here](https://github.com/aiondemand/AIOD-rest-api/commit/bbbb8e34193c47c558b851861c343b09b8045f2b)

File: src/database/model/educational_resource.py (hunk 1 above)
Comment (antonis96) [reply src/database/model/educational_resource.py:68]: To be honest, i am not exactly sure about this.

My suggestion would be the following:
Instead of having only `modified_date` which at the moment is a bit ambiguous, we could have 2 fields (`created_date` and `modified_date`). 
`created_date` will be aligned with the metadata in the corresponding platform while `modified_date` will denote when an update occurred in our metadata storage.

When things get more clear we can revisit this and make any appropriate changes.

File: src/database/model/educational_resource.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (gmouchakis) [reply src/database/model/educational_resource.py]: @josvandervelde our approach is to try to merge resources from the [AIOD catalogue API](https://github.com/aiondemand/AIOD-catalogue-rest-api) with this API. The catalogue resources, on which the CMS was based on, are defined in the "CMS spreadsheet" along with some CMS specific fields. Now that work has begun on the metadata schema as well, we are trying to merge similar fields and include new additions from the metadata schema.

>  It is not ideal, I think, to have 2 sources of truth. Should the Metadata Spreadsheet be updated?
You are right,  we have notified the metadata schema group (Cc @antoniskappa ) about these changes and they will be included in the next version of the metadata schema. I suppose we could merge this now and if the metadata group feels that some fields require modifications we can work on them in a separate issue/PR.

File: src/database/model/educational_resource.py (hunk 1 above)
Comment (gmouchakis) [reply src/database/model/educational_resource.py:68]: > So the question is --> is the metadata storage just a copy of whatever is in other platforms (possibly including an AIoD platform)? Or can there be content that only exists in our metadata storage? I was under impression that it's the former case. What do you think?

From my understating, content related to the AOID catalogue will only exist in the metadata storage. I'm not sure either though how Zenodo related metadata will be handled.

File: src/database/model/educational_resource.py (hunk 11 above)
[the remark this answers is not in the document]
Comment (josvandervelde) [reply src/database/model/educational_resource.py]: Alright, sounds good!

File: src/database/model/educational_resource.py (hunk 1 above)
Comment (josvandervelde) [reply src/database/model/educational_resource.py:68]: Ok, good discussion! I think that having `creation_date` is good, but it wouldn't help here: a resource in for instance Zenodo will also have a date that it was created, and a date in which it was updated.

So let's assume we will have some data that only exists in AIoD, and some that also exist on platforms outside (such as Zenodo). In that case it's fine for me to leave the `server_default`!

File: src/schemas.py (hunk 13 above)
[the remark this answers is not in the document]
Comment (josvandervelde) [reply src/schemas.py]: To be honest, I'm not a big fan of those fields. By just reading the field names, the meaning is unclear to me. But we might want to discuss this indeed in the metadata group, not here. So for me it's fine to leave it for now as is

File: src/converters/educational_resource_converter.py [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+            hours_per_week=aiod.hours_per_week,
+            country=aiod.country,
+            is_accessible_for_free=aiod.is_accessible_for_free,
+            duration_in_years=aiod.duration_in_years,
+            pace=aiod.pace,
+            time_required=aiod.time_required,
+            keywords=[
+                OrmKeyword.as_unique(session=session, name=keyword) for keyword in aiod.keywords
+            ]
+            if aiod.keywords
```
Comment (josvandervelde) [inline src/converters/educational_resource_converter.py:55]: I think it's easier to always have a list, possibly with 0 values, instead of an `Optional[List]`. 
Now you have to handle the `None` case everywhere. You could just use a `default_factory=list` on the `schema.org`

This same point holds also for `business_categories` etc.

File: src/main.py [hunk 15]
```diff
@@ -14,13 +14,14 @@
...
-from sqlalchemy import select, Engine
+from sqlalchemy import Engine
 
 import connectors
 import routers
 from authentication import get_current_user
 from connectors import NodeName
-from database.model.news import OrmNews
+
+# from database.model.base import OrmLanguage, OrmTargetAudience
```
Comment (josvandervelde) [inline src/main.py:24]: This should be removed, I guess?

**Review by josvandervelde — CHANGES_REQUESTED** [review]
Nice! Just 2 minor points

File: src/main.py (hunk 15 above)
Comment (antonis96) [reply src/main.py]: fixed [here](https://github.com/aiondemand/AIOD-rest-api/commit/b94f8496d905919e3d9eb1bb39f88a24fc8db2b6)
